### PR TITLE
8314491: Linux: jexec launched via PATH fails to find java

### PR DESCRIPTION
**Pull request by vpa1977**
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314491](https://bugs.openjdk.org/browse/JDK-8314491): Linux: jexec launched via PATH fails to find java (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15343/head:pull/15343` \
`$ git checkout pull/15343`

Update a local copy of the PR: \
`$ git checkout pull/15343` \
`$ git pull https://git.openjdk.org/jdk.git pull/15343/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15343`

View PR using the GUI difftool: \
`$ git pr show -t 15343`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15343.diff">https://git.openjdk.org/jdk/pull/15343.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15343#issuecomment-1685391697)

### CODE REVIEW

**Comment by vpa1977** [comment]
Running updated test without fix in jexec.c:

```
$make test CONF=linux-x86_64-server-slowdebug TEST=test/jdk/tools/launcher/Jexec.java 
.....

++++Test Output++++
  can't locate java: No such file or directory
++++Test Stack Trace++++
....
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
>> jtreg:test/jdk/tools/launcher/Jexec.java              1     0     1     0 <<
==============================
TEST FAILURE
```

Running test after fixing jexec.c:

```
$ make test CONF=linux-x86_64-server-slowdebug TEST=test/jdk/tools/launcher/Jexec.java 
....
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/jdk/tools/launcher/Jexec.java              1     1     0     0   
==============================
TEST SUCCESS
```

Rerun jdk_launcher tests:
```
$ make test CONF=linux-x86_64-server-slowdebug TEST=jdk_launcher
...
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/jdk:jdk_launcher                         124   124     0     0   
==============================
TEST SUCCESS
```

File: src/java.base/unix/native/launcher/jexec.c
Comment (dholmes-ora) [inline src/java.base/unix/native/launcher/jexec.c:2]: Copyright format is incorrect, please just replace 2017 with 2023.

**Review by dholmes-ora — COMMENTED** [review]
I get the sense from the comment in jexec.c that it is only intended to be launched via a full path, so having it in the $PATH seems like a usage error to me.

That said this change seems relatively harmless, though I confess I do not understand how `/proc/self/exe` works and under what conditions reading it returns the actual binary path rather than "/proc/self/exe". ??

**Comment by vpa1977** [comment]
> I get the sense from the comment in jexec.c that it is only intended to be launched via a full path, so having it in the $PATH seems like a usage error to me.
Unfortunately this executable is linked in /usr/bin and is assumed to work perfectly when run as just `jexec` .

> That said this change seems relatively harmless, though I confess I do not understand how `/proc/self/exe` works and under what conditions reading it returns the actual binary path rather than "/proc/self/exe". ??

The '/proc/' filesystem contains a 'self' entry for the current process and 'exe' entry is expected to contain the symbolic link to the actual binary[1] . `realpath` call in getJavaPath() function translates the symbolic link into the binary path. 

I do not have much expertise in the kernel development, but as far as I know the value of `/proc/self/exe` it is being set in Memory Descriptor structure whenever as process is created from a file[2].
Reading the symlink may return ENOENT (broken symlink) if there was no file[3]

[1] https://docs.kernel.org/filesystems/proc.html
[2] https://github.com/torvalds/linux/blob/f7757129e3dea336c407551c98f50057c22bb266/fs/exec.c#L1282
[3] https://github.com/torvalds/linux/blob/f7757129e3dea336c407551c98f50057c22bb266/fs/proc/base.c#L1730

**Comment by dholmes-ora** [comment]
> realpath call in getJavaPath() function translates the symbolic link into the binary path.

Okay so a specific API makes this work as intended.

What I can't determine is whether this may have an adverse effect on anyone using an "unusual" path to jexec, that `/proc/self/exe` will report differently. To minimise the impact on existing users can this be implemented as a fallback if the initial attempt to locate `java` fails?

**Comment by vpa1977** [comment]
> To minimise the impact on existing users can this be implemented as a fallback if the initial attempt to locate `java` fails?

Updated and re-run the affected test:
```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/jdk/tools/launcher/Jexec.java              1     1     0     0   
==============================
TEST SUCCESS
```

**Review by dholmes-ora — APPROVED** [review]
Thanks for the update, I feel much more comfortable with this change now.

Please wait for a second review however.

Thanks

File: test/jdk/tools/launcher/Jexec.java
Comment (RogerRiggs) [inline test/jdk/tools/launcher/Jexec.java:44]: Check if the imports are really used.  I can't imagine why java.beans.Transient would be needed, ditto UUID, etc.

File: src/java.base/unix/native/launcher/jexec.c
Comment (dholmes-ora) [reply src/java.base/unix/native/launcher/jexec.c:162]: The others are more widely used and there is a desire to document their meaning. That isn't necessary for this case, but as I said, just a nit.

Comment (vpa1977) [reply src/java.base/unix/native/launcher/jexec.c]: Updated and re-run Jexec test. 

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/jdk/tools/launcher/Jexec.java              1     1     0     0   
==============================
TEST SUCCESS
```

**Review by dholmes-ora — APPROVED** [review]
Still good. Thanks

**Comment by vpa1977** [comment]
Would it be ok to type "integrate" at this stage?

**Comment by TheShermanTanker** [comment]
Is this all clear for sponsorship?

**Comment by vpa1977** [comment]
> Is this all clear for sponsorship?

Yes, I have checked with @dholmes-ora

**Review by RogerRiggs — APPROVED** [review]
LGTM

**Pull request closed without merging**
